File: platform/windows/MihomoMeter.Windows.Tests/SharedCoreRuntimeProbeTests.cs
using Microsoft.VisualStudio.TestTools.UnitTesting;
using MihomoMeter.Windows.Core.Application;

namespace MihomoMeter.Windows.Tests;

[TestClass]
public sealed class SharedCoreRuntimeProbeTests
{
    [TestMethod]
    public void ProductionRuntimeProbeLoadsSharedCore()
    {
        Assert.AreEqual(SharedCoreRuntimeStatus.Ready, SharedCoreRuntimeProbe.Run());
    }

    [TestMethod]
    public void RuntimeProbeStopsBeforeNativeCallForAbiMismatch()
    {
        var didScaleTraffic = false;

        var status = SharedCoreRuntimeProbe.Run(
            () => 2,
            _ =>
            {
                didScaleTraffic = true;
                return new SharedTrafficScale(
                    1.5,
                    SharedTrafficUnit.Kilobytes,
                    2);
            });

        Assert.AreEqual(SharedCoreRuntimeStatus.AbiMismatch, status);
        Assert.IsFalse(didScaleTraffic);
    }

    [TestMethod]
    public void RuntimeProbeContainsNativeLoadFailure()
    {
        var status = SharedCoreRuntimeProbe.Run(
            () => throw new DllNotFoundException(),
            _ => throw new AssertFailedException("不应调用缩放函数。"));

        Assert.AreEqual(SharedCoreRuntimeStatus.NativeCallFailed, status);
    }

    [TestMethod]
    public void RuntimeProbeRejectsUnexpectedResult()
    {
        var status = SharedCoreRuntimeProbe.Run(
            () => 1,
            _ => new SharedTrafficScale(
                1.5,
                SharedTrafficUnit.Megabytes,
                2));

        Assert.AreEqual(SharedCoreRuntimeStatus.UnexpectedResult, status);
    }
}

File: Tests/SharedCoreRuntimeProbeTests.swift
import XCTest

@testable import MihomoMeter

final class SharedCoreRuntimeProbeTests: XCTestCase {
  func testProductionRuntimeProbeLoadsSharedCore() {
    XCTAssertEqual(SharedCoreRuntimeProbe.run(), .ready)
  }

  func testRuntimeProbeStopsBeforeNativeCallForABIMismatch() {
    var didScaleTraffic = false

    let status = SharedCoreRuntimeProbe.run(
      abiVersion: { 2 },
      scaleTraffic: { _ in
        didScaleTraffic = true
        return SharedTrafficScale(
          value: 1.5,
          unit: .kilobytes,
          decimalPlaces: 2
        )
      }
    )

    XCTAssertEqual(status, .abiMismatch)
    XCTAssertFalse(didScaleTraffic)
  }

  func testRuntimeProbeMapsAdapterFailureWithoutThrowing() {
    let status = SharedCoreRuntimeProbe.run(
      abiVersion: { 1 },
      scaleTraffic: { _ in
        throw SharedCoreAdapterError.nativeCallFailed(-1)
      }
    )

    XCTAssertEqual(status, .nativeCallFailed)
  }

  func testRuntimeProbeRejectsUnexpectedResult() {
    let status = SharedCoreRuntimeProbe.run(
      abiVersion: { 1 },
      scaleTraffic: { _ in
        SharedTrafficScale(
          value: 1.5,
          unit: .megabytes,
          decimalPlaces: 2
        )
      }
    )

    XCTAssertEqual(status, .unexpectedResult)
  }
}
